### PR TITLE
Proxy settings update

### PR DESCRIPTION
**Pull request by matthew-hagemann**
The Docker snap did not have an easy way to enable us to pass proxy settings to the docker daemon before starting. This prevented us from having a simple lifecycle to manage operating the charm from behind a proxy.

This PR aims to:
- Install docker via a snap, not a deb (snap does not read from `/etc/docker/daemon.json`)
- Remove unused juju library code now that we have swapped from Snaps to debs
- Write proxy settings to `/etc/docker/daemon.json` before docker daemon is started

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -11,7 +11,7 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v4
       - name: Install dependencies
-        run: python3 -m pip install tox
+        run: python3 -m pip install tox --break-system-packages tox
```
Comment (sminez) [inline .github/workflows/build.yaml:14]: what's the need for `--break-system-packages` now?

Comment (matthew-hagemann) [reply .github/workflows/build.yaml:14]: I started running into issues after upping from 22.04 to 24.04 on the gh runners:

https://github.com/canonical/container-runner-operator/actions/runs/11068460084/job/30753946445

Comment (matthew-hagemann) [reply .github/workflows/build.yaml:14]: Probably should have been its own pr :grimacing:

File: src/charm.py [hunk 2]
```diff
@@ -249,10 +254,24 @@ def _db_connection_string(self) -> str:
 
     def _set_proxy(self):
         """Set Squid proxy environment variables if configured."""
-        proxy_url = os.environ.get("JUJU_CHARM_HTTP_PROXY")
-        if proxy_url:
-            os.environ["HTTP_PROXY"] = proxy_url
-            os.environ["HTTPS_PROXY"] = proxy_url
+        http_proxy_env = os.environ.get("JUJU_CHARM_HTTP_PROXY")
```
Comment (sminez) [inline src/charm.py:257]: `get` supports a default value if the key is missing so you could combine these lines with the ones on 263, 264 as:
```python
http_proxy_str = os.environ.get("JUJU_CHARM_HTTP_PROXY", "")
```

File: src/container_runner.py [hunk 3]
```diff
@@ -1,38 +1,42 @@
...
-        return cache["docker"]
-
     def install(self):
-        # Install Docker
         try:
-            self._docker.ensure(snap.SnapState.Latest, channel="stable")
-            snap.hold_refresh()
+            # Set this env to automatically restart daemons. If not, the prompt is blocking.
+            env = os.environ.copy()
+            env["NEEDRESTART_MODE"] = "a"
```
Comment (sminez) [inline src/container_runner.py:28]: might be worth linking to the man page on needrestart to explain what the `"a"` value here is doing: https://manpages.ubuntu.com/manpages/jammy/man1/needrestart.1.html

File: src/container_runner.py [hunk 4]
```diff
@@ -152,6 +156,23 @@ def __init__(self, container_image: str, container_port: int, host_port: int):
         self._container_port = container_port
         self._host_port = host_port
 
+    def set_docker_proxy(self, http_proxy: str, https_proxy: str):
```
Comment (sminez) [inline src/container_runner.py:159]: Like we discussed on our call, the logic here should be inlined into `_set_proxy` and the whole thing should be lifted out as a top level function as `self` isn't used anywhere.

That should also let you inline building out the proxy config which should reduce the line count a bit :slightly_smiling_face:

**Review by sminez — CHANGES_REQUESTED** [review]
looking good :+1: 

just a couple of changes as called out in comments

File: src/container_runner.py [hunk 5]
```diff
@@ -18,6 +18,32 @@
 DOCKER_DAEMON_CONFIG_PATH = Path("/etc/docker/daemon.json")
 
 
+def _try_set_proxy_settings():
+    """If Juju proxy environment variables are present, set procy environment variables and write Docker proxy settings to /etc/docker/daemon.json."""
```
Comment (sminez) [inline src/container_runner.py:22]: typo "procy"

**Review by sminez — APPROVED** [review]
Looks good! Just a typo in one of the doc comments but that's not really a blocker for merging

File: src/container_runner.py (hunk 5 above)
Comment (matthew-hagemann) [reply src/container_runner.py]: Ah, thanks! I'll fix that before I merge :slightly_smiling_face:

**Review by sminez — APPROVED** [review]

**Pull request merged**